resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesClusterSchedulerBackend.scala

```diff
@@ -0,0 +1,456 @@
+      // inform the parent class with removeExecutor.
+      disconnectedPodsByExecutorIdPendingRemoval.keys().asScala.foreach { case (executorId) =>
+        val executorPod = disconnectedPodsByExecutorIdPendingRemoval.get(executorId)
+        val knownExitReason = Option(podsWithKnownExitReasons.remove(
+          executorPod.getMetadata.getName))
+        knownExitReason.fold {
+          removeExecutorOrIncrementLossReasonCheckCount(executorId)
+        } { executorExited =>
+          logDebug(s"Removing executor $executorId with loss reason " + executorExited.message)
+          removeExecutor(executorId, executorExited)
+          // We keep around executors that have exit conditions caused by the application. This
+          // allows them to be debugged later on. Otherwise, mark them as to be deleted from the
+          // the API server.
+          if (!executorExited.exitCausedByApp) {
+            deleteExecutorFromClusterAndDataStructures(executorId)
+          }
+        }
+      }
+    }
+
+    def removeExecutorOrIncrementLossReasonCheckCount(executorId: String): Unit = {
+      val reasonCheckCount = executorReasonCheckAttemptCounts.getOrElse(executorId, 0)
+      if (reasonCheckCount >= MAX_EXECUTOR_LOST_REASON_CHECKS) {
+        removeExecutor(executorId, SlaveLost("Executor lost for unknown reasons."))
+        deleteExecutorFromClusterAndDataStructures(executorId)
+      } else {
+        executorReasonCheckAttemptCounts.put(executorId, reasonCheckCount + 1)
+      }
+    }
+
+    def deleteExecutorFromClusterAndDataStructures(executorId: String): Unit = {
+      disconnectedPodsByExecutorIdPendingRemoval.remove(executorId)
+      executorReasonCheckAttemptCounts -= executorId
+      RUNNING_EXECUTOR_PODS_LOCK.synchronized {
+        runningExecutorsToPods.remove(executorId).map { pod =>
+          kubernetesClient.pods().delete(pod)
+          runningPodsToExecutors.remove(pod.getMetadata.getName)
+        }.getOrElse(logWarning(s"Unable to remove pod for unknown executor $executorId"))
+      }
+    }
+  }
+
+  private def getInitialTargetExecutorNumber(defaultNumExecutors: Int = 1): Int = {
+    if (Utils.isDynamicAllocationEnabled(conf)) {
+      val minNumExecutors = conf.getInt("spark.dynamicAllocation.minExecutors", 0)
+      val initialNumExecutors = Utils.getDynamicAllocationInitialExecutors(conf)
+      val maxNumExecutors = conf.getInt("spark.dynamicAllocation.maxExecutors", 1)
+      require(initialNumExecutors >= minNumExecutors && initialNumExecutors <= maxNumExecutors,
+        s"initial executor number $initialNumExecutors must between min executor number " +
+          s"$minNumExecutors and max executor number $maxNumExecutors")
+
+      initialNumExecutors
+    } else {
+      conf.getInt("spark.executor.instances", defaultNumExecutors)
+    }
+
+  }
+
+  override def applicationId(): String = conf.get("spark.app.id", super.applicationId())
```

Comment:
Why is this method necessary?

Reply:
We use the application ID for labeling the driver and executors pods in k8s so users can aggregate and filter by that label to obtain all execution units belonging to a particular job.

Reply:
...execution units belonging to a particular Spark Application.

not job

Reply:
Noted, thanks! (I meant "job" in the context of a "batch/streaming job", but I understand that's an overloaded term wrt Spark.)

Reply:
Sorry, that doesn't answer my question. Are you saying that you let users define the application ID?

That would be a bad idea since Spark assumes those are unique in certain places (e.g. the event logger and the history server).

Reply:
I think this is an artifact of missing the changes to `spark-submit` that are specific to Kubernetes. The Kubernetes implementation of `spark-submit` that we intend to ship generates a unique value for `spark.app.id`, sets it as a JVM option for the driver, and wires that value into other Kubernetes resources that are tied to the application.

Reply:
But since we didn't include the spark-submit implementation in this PR, this implementation of `applicationId()` lacks context.

Reply:
> generates a unique value for spark.app.id, sets it as a JVM option for the driver

Why can't the id generation be done here, like it's done in all other cluster managers? `SparkContext` already takes care of setting that value in the `conf` object used internally.

Reply:
It can't be done here because `spark-submit` would like to know this shared ID a-priori before the driver pod even comes into existence.

Reply:
Although `spark-submit` could have its own "kubernetes Spark ID" which is different from this app ID, but it seems confusing to have multiple representations of a Spark application's identifier.

Reply:
Is there anywhere where these spark-submit changes you mention can be seen? You're starting to get me worried about what you guys are doing.

spark-submit is supposed to be dumb. It just runs a class; in client mode that's the user's main class, which should eventually create a `SparkContext` and trigger this code.

In cluster mode it's a cluster-manager specific class that submits the user application to run in the cluster somewhere.

I can't see why you'd need a pre-computed app id in either case unless you're trying to make spark-submit do much more than it should.

Reply:
Spark-submit is actually relatively complex in Kubernetes mode and we intend to propose it in chunks. We describe the overall architecture here: https://github.com/apache-spark-on-k8s/spark/blob/branch-2.2-kubernetes/resource-managers/kubernetes/architecture-docs/submission-client.md

Implementation at the top level is here: https://github.com/apache-spark-on-k8s/spark/blob/branch-2.2-kubernetes/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/Client.scala

Reply:
To clarify, spark-submit in the above conversation refers to the client-side component of K8s mode, and not actually SparkSubmit in spark-core.

Reply:
The spark-submit code is pretty standard - it's simply running the [Client](https://github.com/apache-spark-on-k8s/spark/blob/branch-2.2-kubernetes/core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala#L636) as one might expect.  

@vanzin, just to summarize - In our current setup, we are computing the `spark.app.id` in our own Client and passing it on to the `KubernetesClusterSchedulerBackend`. We're not letting the user modify the `spark.app.id` but [setting it ourselves](https://github.com/apache-spark-on-k8s/spark/blob/branch-2.2-kubernetes/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/Client.scala#L160).

Reply:
>  it's simply running the Client as one might expect.

Yeah, but what about client mode? In client mode the first time the k8s code will be invoked is when the scheduler backend is called. So you can't rely on anything else setting `spark.app.id` for you. In that case here you'd end up with whatever the super class defines - unless the user sets an app id themselves, and that shouldn't be allowed.

Reply:
Client mode on k8s will look very different from that on other cluster managers. This is because in almost all k8s clusters, networking in setup such that pods (or simply containers) within clusters are not routable directly from outside, and vice versa. 

An alternative to support the same functionality is ["in-cluster client mode"](https://github.com/apache-spark-on-k8s/spark/pull/456). This will look similar to cluster mode (driver and executor will run within the cluster), except that we'd set a couple of additional parameters on the driver pod, to allow someone outside the cluster to connect with it interactively (using the Kubernetes APIServer as a proxy). We don't foresee a design in which the `KubernetesClusterSchedulerBackend` will be invoked directly on the client-side, because there is no standard mechanism by which executors can reach back out to the driver in k8s (this is by design, and the reasoning is explained in part by the requirements in https://kubernetes.io/docs/concepts/cluster-administration/networking/#kubernetes-model).

Reply:
Ok, this is starting to sound more like a review of your spark-submit integration. Can you move this method implementation to that PR later on, since then the git history will actually give you more context about why this particular implementation is needed?

Reply:
@vanzin, your general concern is understood. I think here the key difference from YARN is that we don't have the notion of asking k8s for a unique ID and "registering the application" in the same way which is why we need to pass this ID to the components we create - the driver and the executors. 

Like @mccheah said, we could actually avoid using `spark.app.id` to identify and name various components we create in the cluster, and just use our a k8s-specific random generated ID instead, but that might make it difficult for people to reconcile what they see in the history server or other UIs with what they see in Kubernetes (driver & executor pods). If you feel strongly about the setting of `spark.app.id`, we can do this. So far from usage we've seen, I don't think it has been surprising to people. 

If you don't feel strongly about it, we can document this discussion with a code comment for now - so we don't lose this thread which has important concerns - and if we get to a point where it becomes evident that it's causing issues/confusion, we can move to the above approach.

Reply:
> Ok, this is starting to sound more like a review of your spark-submit integration. Can you move this method implementation to that PR later on, since then the git history will actually give you more context about why this particular implementation is needed?
 
Yeah, this SGTM. We can do that and revisit this later.